pkg/p2p/kadcli/reader.go

```diff
@@ -0,0 +1,121 @@
+// This Source Code Form is subject to the terms of the MIT License.
+// If a copy of the MIT License was not distributed with this
+// file, you can obtain one at https://opensource.org/licenses/MIT.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+package kadcli
+
+import (
+	"bytes"
+	"context"
+	"encoding/hex"
+	"errors"
+	"io"
+
+	"github.com/dusk-network/dusk-blockchain/pkg/p2p/peer"
+	"github.com/dusk-network/dusk-blockchain/pkg/p2p/wire/checksum"
+	"github.com/dusk-network/dusk-blockchain/pkg/p2p/wire/protocol"
+	"github.com/dusk-network/dusk-blockchain/pkg/p2p/wire/topics"
+	"github.com/dusk-network/dusk-blockchain/pkg/util/nativeutils/eventbus"
+	"github.com/dusk-network/dusk-protobuf/autogen/go/rusk"
+)
+
+// Reader abstracts all of the logic and fields needed to receive messages from
+// other network nodes.
+type Reader struct {
+	processor *peer.MessageProcessor
+	gossip    *protocol.Gossip
+
+	client rusk.NetworkClient
+	stop   context.CancelFunc
+}
+
+// NewReader makes a new kadcast reader that handles TCP packets of broadcasting.
+func NewReader(publisher eventbus.Publisher, g *protocol.Gossip, p *peer.MessageProcessor, rusk rusk.NetworkClient) *Reader {
+	return &Reader{
+		processor: p,
+		gossip:    g,
+		client:    rusk,
+	}
+}
+
+// Listen starts accepting and processing stream data.
+func (r *Reader) Listen() {
+	// create stream handler
+	stream, err := r.client.Listen(context.Background(), &rusk.Null{})
+	if err != nil {
+		log.WithError(err).Fatal("open stream error")
+		return
+	}
+
+	// create a context
+	ctx, cancel := context.WithCancel(context.Background())
+	r.stop = cancel
+
+	// listen for messages
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+				// receive a message
+				msg, err := stream.Recv()
+				if err == io.EOF {
+					log.Warnln("reader reached EOF")
+					return
+				} else if err != nil {
+					log.WithError(err).Fatal("receive error")
+				}
+				// Message received
+				go r.processMessage(msg)
+			}
+		}
+	}()
+}
+
+// processMessage propagates the received kadcast message into the event bus.
+func (r *Reader) processMessage(message *rusk.Message) {
+	reader := bytes.NewReader(message.Message)
+
+	// read message (extract length and magic)
+	b, err := r.gossip.ReadMessage(reader)
+	if err != nil {
+		log.WithField("r_addr", message.Metadata.SrcAddress).
+			WithError(err).Warnln("error reading message")
+		return
+	}
+	// extract checksum
+	msg, cs, err := checksum.Extract(b)
+	if err != nil {
+		log.WithError(err).Warnln("error extracting message and cs")
+		return
+	}
+	// verify checksum
+	if !checksum.Verify(msg, cs) {
+		log.WithError(errors.New("invalid checksum")).
+			Warnln("error verifying message cs")
+		return
+	}
+
+	// collect (process) the message
+	go func() {
+		if _, err = r.processor.Collect(message.Metadata.SrcAddress, msg, nil, protocol.FullNode, nil); err != nil {
```

Comment:
Keep in mind that we need to turn the responses into `KadcastPoint` messages otherwise syncing will not work.

```golang

// TODO: set service flag properly
	respBufs, err := r.processor.Collect(raddr, m, nil, protocol.FullNode, []byte{height})
	if err != nil {
		ll.WithError(err).Error("messageProcessor failed to collect message")
	}

	// Any response message is translated into Kadcast Point-to-Point wire message.
	// In other words, any bufs item is sent back to the sender node (remotePeer).
	for i := 0; i < len(respBufs); i++ {
		ll.WithField("raddr", remotePeer.Address()).Trace("send point-to-point message")

		// Send Kadcast point-to-point message with destination raddr
		r.publisher.Publish(topics.KadcastPoint,
			message.NewWithHeader(topics.KadcastPoint, respBufs[i], []byte(remotePeer.Address())))
	}
```